_posts/2023-10-04-Data-Quality-Checks.md

```diff
@@ -0,0 +1,408 @@
+Hoare logic includes the idea of a [Hoare triple](https://en.wikipedia.org/wiki/Hoare_logic#Hoare_triple) ($ {\displaystyle \{P\}C\{Q\}} $) where $ {\displaystyle \{P\}} $ is an assertion of precondition, $ {\displaystyle \ C} $ is a command, and $ {\displaystyle \{Q\}} $ is a postcondition assertion.
+Software development using data often entails (sometimes assumed) assertions of precondition from data sources, a transformation or command which changes the data, and a (sometimes assumed) assertion of postcondition in a data output or result.
+
+### Design by Contract
+
+```mermaid!
+flowchart LR
+    subgraph local ["Data Testing through Design by Contract over Hoare Triple"]
+        direction LR
+        subgraph hoare_triple ["Hoare Triple"]
+            direction LR
+            input_data[("Input Data\n(P - precondition)")]
+            
+            process_data["Data processing\n(C - command)"]
+            
+            output_data[("Output Data\n(Q - postcondition)")]
+        end
+        subgraph dbc ["Design by Contract"]
+            direction LR
+            met_specification1{"Met\nspecs?"}
+            contract1(["Contract(s)"])
+            met_specification2{"Met \nspecs?"}
+            contract2(["Contract(s)"])
+        end
+    end
+
+    input_data --> met_specification1
+    met_specification1 --> process_data
+    process_data --> met_specification2
+    met_specification2 --> output_data
+    contract1 --> met_specification1
+    contract2 --> met_specification2
+
+```
+
+Hoare logic and Software correctness help describe [Design by contract (DbC)](https://en.wikipedia.org/wiki/Design_by_contract), a software approach involving the formal specification of "contracts" which help ensure we meet our intended goals.
+DbC helps describe how to create assertions when proceeding through Hoare triplet states for data.
+These concepts provide a framework for thinking about the tools mentioned below.
+
+## Data Component Testing
+
+```mermaid!
+flowchart LR
+    input_data[("Data to check")]
+    has_components{"Exhibits\ncomponents?"}
+    contract(["Contract(s)"])
+    components["- Has no null values\n- Columns are all numeric\n- etc..."]
+    continue["Continue\noperations"]
+    error["Raise\nexception"]
+
+    components --> contract
+    input_data --> has_components
+    contract -.-> has_components
+    has_components --> | Yes | continue
+    has_components --> | No | error
+
+    style error fill:#FED7AA
+    style continue fill:#DCFCE7
+    style components fill:#fff
+```
+
+_Diagram showing data contracts as __generalized and reusable "component" testing__ being checked through contracts and raising an error if they aren't met or continuing operations if they are met._
+
+We often need to verify a certain components surrounding data in order to ensure it meets minimum standards.
```

Comment:
I assumed you were talking about the data surrounding a certain component, so I added an apostrophe to make it possessive. It's less likely IMO, but if you instead meant multiple components that surround data, you'd drop the article ("a") instead.

```suggestion
We often need to verify a certain component's surrounding data in order to ensure it meets minimum standards.
```

Reply:
Ah, great catch, thanks! Adding this in.